From 93fb2f3d393c69471ec454491d18e4ce3cdc083e Mon Sep 17 00:00:00 2001
From: Akshay Adhikari <aadhikar@redhat.com>
Date: Tue, 18 Nov 2025 21:57:10 +0530
Subject: [PATCH] Issue 7076, 6992, 6784, 6214 - Fix CI test failures (#7077)

- Fixed ModRDN cache corruption on failed operations (parent update check, cache cleanup)
- Fixed attribute uniqueness test fixture cleanup in attruniq_test.py
- Improved test stability by fixing race conditions in replication,
  web UI, and basic tests.

Fixes: https://github.com/389ds/389-ds-base/issues/7076
Relates: https://github.com/389ds/389-ds-base/issues/6214

Reviewed by: @vashirov, @progier389 (Thanks!)
---
 dirsrvtests/tests/suites/basic/basic_test.py  |  4 +-
 .../tests/suites/plugins/attruniq_test.py     | 86 +++++++++++--------
 .../suites/replication/regression_m2_test.py  |  4 +
 .../suites/webui/database/database_test.py    |  1 +
 ldap/servers/slapd/back-ldbm/ldbm_modrdn.c    | 44 ++++++----
 5 files changed, 83 insertions(+), 56 deletions(-)

diff --git a/dirsrvtests/tests/suites/basic/basic_test.py b/dirsrvtests/tests/suites/basic/basic_test.py
index be825efe9..e9b611439 100644
--- a/dirsrvtests/tests/suites/basic/basic_test.py
+++ b/dirsrvtests/tests/suites/basic/basic_test.py
@@ -593,7 +593,7 @@ def test_basic_import_export(topology_st, import_example_ldif):
     #
     # Test online/offline LDIF imports
     #
-    topology_st.standalone.start()
+    topology_st.standalone.restart()
     # topology_st.standalone.config.set('nsslapd-errorlog-level', '1')
 
     # Generate a test ldif (50k entries)
@@ -691,6 +691,8 @@ def test_basic_backup(topology_st, import_example_ldif):
 
     log.info('Running test_basic_backup...')
 
+    topology_st.standalone.restart()
+
     backup_dir = topology_st.standalone.get_bak_dir() + '/backup_test_online'
     log.info(f'Backup directory is {backup_dir}')
 
diff --git a/dirsrvtests/tests/suites/plugins/attruniq_test.py b/dirsrvtests/tests/suites/plugins/attruniq_test.py
index ab07e37e2..a2be413c8 100644
--- a/dirsrvtests/tests/suites/plugins/attruniq_test.py
+++ b/dirsrvtests/tests/suites/plugins/attruniq_test.py
@@ -27,6 +27,7 @@ pytestmark = [pytest.mark.tier1,
 logging.getLogger(__name__).setLevel(logging.DEBUG)
 log = logging.getLogger(__name__)
 MAIL_ATTR_VALUE = 'non-uniq@value.net'
+MAIL_ATTR_VALUE_ALT = 'alt-mail@value.net'
 
 EXCLUDED_CONTAINER_CN = "excluded_container"
 EXCLUDED_CONTAINER_DN = "cn={},{}".format(EXCLUDED_CONTAINER_CN, DEFAULT_SUFFIX)
@@ -83,14 +84,23 @@ def containers(topology_st, request):
 def attruniq(topology_st, request):
     log.info('Setup attribute uniqueness plugin')
     attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
-    attruniq.create(properties={'cn': 'attruniq'})
-    attruniq.add_unique_attribute('cn')
+
+    if attruniq.exists():
+        attruniq.delete()
+        topology_st.standalone.restart()
+
+    attruniq.create(properties={
+        'cn': 'attruniq',
+        'uniqueness-attribute-name': 'cn',
+        'uniqueness-subtrees': 'cn=config',
+        'nsslapd-pluginEnabled': 'on'
+    })
     topology_st.standalone.restart()
 
     def fin():
         if attruniq.exists():
-            attruniq.disable()
             attruniq.delete()
+            topology_st.standalone.restart()
 
     request.addfinalizer(fin)
 
@@ -249,8 +259,8 @@ def test_modrdn_attr_uniqueness(topology_st, attruniq):
     group1 = groups.create(properties={'cn': 'group1'})
     group2 = groups.create(properties={'cn': 'group2'})
 
-    attruniq.add_unique_attribute('mail')
-    attruniq.add_unique_subtree(group2.dn)
+    attruniq.replace('uniqueness-attribute-name', 'mail')
+    attruniq.replace('uniqueness-subtrees', group2.dn)
     attruniq.enable_all_subtrees()
     log.debug(f'Enable PLUGIN_ATTR_UNIQUENESS plugin as "ON"')
     attruniq.enable()
@@ -273,10 +283,6 @@ def test_modrdn_attr_uniqueness(topology_st, attruniq):
     assert 'attribute value already exist' in str(excinfo.value)
     log.debug(excinfo.value)
 
-    log.debug('Move user2 to group1')
-
-    user2.rename(f'uid={user2.rdn}', group1.dn)
-
     user1.delete()
     user2.delete()
 
@@ -302,17 +308,11 @@ def test_multiple_attr_uniqueness(topology_st, attruniq):
         6. Should raise CONSTRAINT_VIOLATION
     """
 
-    try:
-        attruniq.add_unique_attribute('mail')
-        attruniq.add_unique_attribute('mailAlternateAddress')
-        attruniq.add_unique_subtree(DEFAULT_SUFFIX)
-        attruniq.enable_all_subtrees()
-        log.debug(f'Enable PLUGIN_ATTR_UNIQUENESS plugin as "ON"')
-        attruniq.enable()
-    except ldap.LDAPError as e:
-        log.fatal('test_multiple_attribute_uniqueness: Failed to configure plugin for "mail": error {}'.format(e.args[0]['desc']))
-        assert False
-
+    attruniq.replace('uniqueness-attribute-name', ['mail', 'mailAlternateAddress'])
+    attruniq.replace('uniqueness-subtrees', DEFAULT_SUFFIX)
+    attruniq.enable_all_subtrees()
+    log.debug(f'Enable PLUGIN_ATTR_UNIQUENESS plugin as "ON"')
+    attruniq.enable()
     topology_st.standalone.restart()
 
     users = UserAccounts(topology_st.standalone, DEFAULT_SUFFIX)
@@ -383,8 +383,9 @@ def test_exclude_subtrees(topology_st, attruniq):
         16. Success
         17. Success
     """
-    attruniq.add_unique_attribute('telephonenumber')
-    attruniq.add_unique_subtree(DEFAULT_SUFFIX)
+    # Replace dummy config with actual test config
+    attruniq.replace('uniqueness-attribute-name', 'telephonenumber')
+    attruniq.replace('uniqueness-subtrees', DEFAULT_SUFFIX)
     attruniq.enable_all_subtrees()
     attruniq.enable()
     topology_st.standalone.restart()
@@ -517,10 +518,18 @@ def test_matchingrule_attr(topology_st):
     """
 
     inst = topology_st.standalone
-
     attruniq = AttributeUniquenessPlugin(inst,
                                          dn="cn=attribute uniqueness,cn=plugins,cn=config")
-    attruniq.add_unique_attribute('cn:CaseExactMatch:')
+
+    if attruniq.exists():
+        attruniq.delete()
+        inst.restart()
+
+    attruniq.create(properties={
+        'cn': 'attribute uniqueness',
+        'uniqueness-attribute-name': 'cn:CaseExactMatch:',
+        'uniqueness-subtrees': DEFAULT_SUFFIX
+    })
     attruniq.enable_all_subtrees()
     attruniq.enable()
     inst.restart()
@@ -595,7 +604,7 @@ def test_one_container_add(topology_st, attruniq, containers, active_user_1):
     active_2.delete()
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.replace('uniqueness-subtrees', ACTIVE_DN)
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -628,7 +637,7 @@ def test_one_container_mod(topology_st, attruniq, containers,
     """
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.replace('uniqueness-subtrees', ACTIVE_DN)
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -656,7 +665,7 @@ def test_one_container_modrdn(topology_st, attruniq, containers,
     """
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
+    attruniq.replace('uniqueness-subtrees', ACTIVE_DN)
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -690,8 +699,7 @@ def test_multiple_containers_add(topology_st, attruniq, containers,
     """
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
-    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.replace('uniqueness-subtrees', [ACTIVE_DN, STAGE_DN])
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -789,8 +797,7 @@ def test_multiple_containers_mod(topology_st, attruniq, containers,
     """
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
-    attruniq.add_unique_subtree(STAGE_DN)
+    attruniq.replace('uniqueness-subtrees', [ACTIVE_DN, STAGE_DN])
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -874,8 +881,8 @@ def test_multiple_containers_modrdn(topology_st, attruniq, containers,
     """
 
     log.info('Setup attribute uniqueness plugin for "cn" attribute')
-    attruniq.add_unique_subtree(ACTIVE_DN)
-    attruniq.add_unique_subtree(STAGE_DN)
+    # Replace dummy subtree with actual test subtrees
+    attruniq.replace('uniqueness-subtrees', [ACTIVE_DN, STAGE_DN])
     attruniq.enable()
     topology_st.standalone.restart()
 
@@ -993,7 +1000,10 @@ def test_invalid_config_missing_attr_name(topology_st):
     _config_file(topology_st, action='save')
 
     attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
-    attruniq.create(properties={'cn': 'attruniq'})
+    attruniq.create(properties={
+        'cn': 'attruniq',
+        'uniqueness-subtrees': DEFAULT_SUFFIX
+    })
     attruniq.enable()
 
     topology_st.standalone.errorlog_file = open(topology_st.standalone.errlog, "r")
@@ -1040,9 +1050,11 @@ def test_invalid_config_invalid_subtree(topology_st):
     _config_file(topology_st, action='save')
 
     attruniq = AttributeUniquenessPlugin(topology_st.standalone, dn="cn=attruniq,cn=plugins,cn=config")
-    attruniq.create(properties={'cn': 'attruniq'})
-    attruniq.add_unique_attribute('cn')
-    attruniq.add_unique_subtree('invalid_subtree')
+    attruniq.create(properties={
+        'cn': 'attruniq',
+        'uniqueness-attribute-name': 'cn',
+        'uniqueness-subtrees': 'invalid_subtree'
+    })
     attruniq.enable()
 
     topology_st.standalone.errorlog_file = open(topology_st.standalone.errlog, "r")
diff --git a/dirsrvtests/tests/suites/replication/regression_m2_test.py b/dirsrvtests/tests/suites/replication/regression_m2_test.py
index ba1ffcc9c..db5140b0b 100644
--- a/dirsrvtests/tests/suites/replication/regression_m2_test.py
+++ b/dirsrvtests/tests/suites/replication/regression_m2_test.py
@@ -1221,6 +1221,10 @@ def test_rid_starting_with_0(topo_m2, request):
     for replica,rid in zip(replicas, ['010', '020']):
         replica.replace('nsDS5ReplicaId', rid)
 
+    # Restart required - replica IDs are loaded at startup and cached in memory
+    S1.restart()
+    S2.restart()
+
     # Restore replica id in finalizer
     def fin():
         for replica,rid in zip(replicas, ['1', '2']):
diff --git a/dirsrvtests/tests/suites/webui/database/database_test.py b/dirsrvtests/tests/suites/webui/database/database_test.py
index ef105d262..05ddb6b00 100644
--- a/dirsrvtests/tests/suites/webui/database/database_test.py
+++ b/dirsrvtests/tests/suites/webui/database/database_test.py
@@ -138,6 +138,7 @@ def test_chaining_configuration_availability(topology_st, page, browser_name):
     log.info('Click on Chaining Configuration and check if element is loaded.')
     frame.get_by_role('tab', name='Database', exact=True).click()
     frame.locator('#chaining-config').click()
+    frame.locator('#chaining-page').wait_for()
     frame.locator('#defSizeLimit').wait_for()
     assert frame.locator('#defSizeLimit').is_visible()
 
diff --git a/ldap/servers/slapd/back-ldbm/ldbm_modrdn.c b/ldap/servers/slapd/back-ldbm/ldbm_modrdn.c
index e3b7e5783..b1a29ff7f 100644
--- a/ldap/servers/slapd/back-ldbm/ldbm_modrdn.c
+++ b/ldap/servers/slapd/back-ldbm/ldbm_modrdn.c
@@ -1080,7 +1080,9 @@ ldbm_back_modrdn(Slapi_PBlock *pb)
                 }
             }
         }
-        if (slapi_sdn_get_dn(dn_newsuperiordn) != NULL) {
+        /* Only update parent if we're actually moving to a NEW parent (not the same parent) */
+        if (slapi_sdn_get_dn(dn_newsuperiordn) != NULL &&
+            slapi_sdn_compare(dn_newsuperiordn, &dn_parentdn) != 0) {
             /* Push out the db modifications from the parent entry */
             retval = modify_update_all(be, pb, &parent_modify_context, &txn);
             if (DBI_RC_RETRY == retval) {
@@ -1335,11 +1337,6 @@ ldbm_back_modrdn(Slapi_PBlock *pb)
     goto common_return;
 
 error_return:
-    /* Revert the caches if this is the parent operation */
-    if (parent_op && betxn_callback_fails) {
-        revert_cache(inst, &parent_time);
-    }
-
     /* result already sent above - just free stuff */
     if (postentry) {
         slapi_entry_free(postentry);
@@ -1417,13 +1414,6 @@ error_return:
                     slapi_pblock_set(pb, SLAPI_PLUGIN_OPRETURN, ldap_result_code ? &ldap_result_code : &retval);
                 }
                 slapi_pblock_get(pb, SLAPI_PB_RESULT_TEXT, &ldap_result_message);
-
-                /* As it is a BETXN plugin failure then
-                 * revert the caches if this is the parent operation
-                 */
-                if (parent_op) {
-                    revert_cache(inst, &parent_time);
-                }
             }
             retval = plugin_call_mmr_plugin_postop(pb, NULL,SLAPI_PLUGIN_BE_TXN_POST_MODRDN_FN);
 
@@ -1437,6 +1427,15 @@ error_return:
         }
     }
 
+    /* Revert the caches if this is the parent operation and cache modifications were made.
+     * Cache modifications (via modify_switch_entries) only happen after BETXN PRE plugins succeed,
+     * so we should only revert if we got past that point (i.e., BETXN POST plugin failures).
+     * For BETXN PRE failures, no cache modifications were made to parent/newparent entries.
+     */
+    if (parent_op && betxn_callback_fails && postentry) {
+        revert_cache(inst, &parent_time);
+    }
+
 common_return:
 
     /* result code could be used in the bepost plugin functions. */
@@ -1482,12 +1481,22 @@ common_return:
                                       "operation failed, the target entry is cleared from dncache (%s)\n", slapi_entry_get_dn(ec->ep_entry));
             CACHE_REMOVE(&inst->inst_dncache, bdn);
             CACHE_RETURN(&inst->inst_dncache, &bdn);
+
+            /* Also remove ec from entry cache and free it since the operation failed */
+            if (inst && cache_is_in_cache(&inst->inst_cache, ec)) {
+                CACHE_REMOVE(&inst->inst_cache, ec);
+                CACHE_RETURN(&inst->inst_cache, &ec);
+            } else {
+                /* ec was not in cache, just free it */
+                backentry_free(&ec);
+            }
+            ec = NULL;
         }
 
         if (ec && inst) {
             CACHE_RETURN(&inst->inst_cache, &ec);
+            ec = NULL;
         }
-        ec = NULL;
     }
 
     if (inst) {
@@ -1817,6 +1826,8 @@ moddn_get_newdn(Slapi_PBlock *pb, Slapi_DN *dn_olddn, Slapi_DN *dn_newrdn, Slapi
 
 /*
  * Return the entries to the cache.
+ * For the original entry 'e', we should NOT remove it from cache on failure,
+ * as it's still a valid entry in the directory.
  */
 static void
 moddn_unlock_and_return_entry(
@@ -1825,12 +1836,9 @@ moddn_unlock_and_return_entry(
 {
     ldbm_instance *inst = (ldbm_instance *)be->be_instance_info;
 
-    /* Something bad happened so we should give back all the entries */
+    /* Unlock and return the entry to the cache */
     if (*targetentry != NULL) {
         cache_unlock_entry(&inst->inst_cache, *targetentry);
-        if (cache_is_in_cache(&inst->inst_cache, *targetentry)) {
-            CACHE_REMOVE(&inst->inst_cache, *targetentry);
-        }
         CACHE_RETURN(&inst->inst_cache, targetentry);
         *targetentry = NULL;
     }
-- 
2.52.0

